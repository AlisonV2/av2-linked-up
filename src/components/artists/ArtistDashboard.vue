<template>
  <div class="row">
    <app-title mode="title-img">Your profile</app-title>
    <div class="toast-success" v-if="showSuccessToast">Update successful</div>
    <div class="toast-error" v-if="showErrorToast">
      Oops... Something went wrong :(
    </div>
    <div class="col-12 col-lg-4 mt-4">
      <div class="mt-3">
        <input
          type="file"
          data-test="file-upload"
          @change="setPreviewImage($event)"
        />
        <div class="error">{{ fileErr }}</div>
      </div>
      <img
        :src="
          previewImage
            ? previewImage
            : require('@/assets/img/default-placeholder.png')
        "
        width="250"
        height="250"
        class="img-fluid"
      />
      <div class="btn-center mt-3 thumbnail-button" @click="setArtistThumbnail">
        <app-button mode="save-btn">Save</app-button>
      </div>
      <!-- <div class="error">{{ fileError }}</div> -->
    </div>
    <form class="col-12 col-lg-8 mt-4" @submit.prevent="setArtistProfile">
      <div class="form-floating mb-3">
        <input
          type="text"
          class="form-control"
          placeholder="Public Name"
          v-model="profile.name"
        />
        <label>Public Name</label>
      </div>
      <div class="form-floating mb-3">
        <input
          type="textarea"
          class="form-control"
          placeholder="Description "
          v-model="profile.description"
        />
        <label>Description</label>
      </div>
      <div class="input-group">
        <div class="form-floating mb-3 col-lg-6">
          <input
            type="text"
            class="form-control"
            placeholder="City"
            v-model="profile.city"
          />
          <label>City</label>
        </div>
        <div class="form-floating mb-3 col-lg-6">
          <input
            type="text"
            class="form-control"
            placeholder="Zip Code"
            v-model="profile.zip"
          />
          <label>Zip Code</label>
        </div>
      </div>
      <div class="form-floating mb-3">
        <input
          type="text"
          class="form-control"
          placeholder="Shop"
          v-model="profile.shop"
        />
        <label>Shop</label>
      </div>
      <div class="form-floating mb-3">
        <input
          type="text"
          class="form-control"
          placeholder="Social Media Link"
          v-model="profile.socialLink"
        />
        <label>Social Media Link</label>
      </div>
      <div class="form-floating mb-3">
        <input
          type="text"
          class="form-control"
          placeholder="Instagram Name"
          v-model="profile.insta"
        />
        <label>Instagram Name</label>
      </div>
      <div>
        <select class="form-select" v-model="profile.style">
          <option disabled value="">Choose your style</option>
          <option value="Black Work">Black Work</option>
          <option value="New School">New School</option>
          <option value="Old School">Old School</option>
          <option value="Surrealism">Surrealism</option>
        </select>
      </div>
      <div class="col-12 btn-right mt-3">
        <div class="btn-group mb-3">
          <app-button>Save</app-button>
        </div>
      </div>
    </form>
  </div>
</template>

<script>
import * as Sentry from '@sentry/vue';

/**
 * @exports ArtistDashboard
 * @type {Component}
 * @requires Sentry
 * @vue-data {array} previewImage
 * @vue-data {object} profile
 * @vue-data {boolean} showSuccessToast
 * @vue-data {boolean} showErrorToast
 * @vue-data {array} fileSelected
 * @vue-data {string} fileErr
 * @vue-event {object} getArtistProfile
 * @vue-event {string} setPreviewImage
 * @vue-event {string} setArtistProfile
 * @vue-event {string} setArtistThumbnail
 */
export default {
  name: 'ArtistDashboard',
  data() {
    return {
      previewImage: null,
      profile: {
        name: '',
        city: '',
        zip: '',
        shop: '',
        style: '',
        socialLink: '',
        insta: '',
        description: '',
        thumbnail: '',
      },
      showSuccessToast: false,
      showErrorToast: false,
      fileSelected: null,
      fileErr: null,
    };
  },
  async created() {
    await this.$store
      .dispatch('getArtistProfile')
      .then(() => {
        this.profile = this.$store.state.artistProfile.artistProfile;
      })
      .then(() => {
        this.previewImage = this.profile.thumbnail;
      })
      .catch((err) => Sentry.captureException(err));
  },
  methods: {
    setPreviewImage($event) {
      const file = $event.target.files[0];
      const types = ['image/png', 'image/jpeg'];

      if (file && types.includes(file.type)) {
        this.fileSelected = file;
        this.fileErr = null;
        this.previewImage = URL.createObjectURL(file);
      } else {
        this.fileSelected = null;
        this.fileErr = 'Please select an image file (png or jpeg).';
      }
    },
    setArtistThumbnail() {
      this.$store
        .dispatch('setArtistThumbnail', this.fileSelected)
        .then(() => {
          this.showSuccess();
          window.location.reload();
        })
        .catch((err) => {
          this.showError(err);
        });
    },
    async setArtistProfile() {
      const profileData = {
        name: this.profile.name,
        city: this.profile.city,
        zip: this.profile.zip,
        shop: this.profile.shop,
        style: this.profile.style,
        socialLink: this.profile.socialLink,
        insta: this.profile.insta,
        description: this.profile.description,
      };
      await this.$store
        .dispatch('setArtistProfile', profileData)
        .then(() => {
          this.showSuccess();
          window.location.reload();
        })
        .catch((err) => {
          this.showError(err);
        });
    },
    showSuccess() {
      this.showSuccessToast = true;
      this.showErrorToast = false;
      setTimeout(() => {
        this.showSuccessToast = false;
      }, 3000);
    },
    showError(err) {
      this.showErrorToast = true;
      this.showSuccessToast = false;
      Sentry.captureException(err);
      setTimeout(() => {
        this.showErrorToast = false;
      }, 3000);
    },
  },
};
</script>

<style lang="scss" scoped>
@include bp-down(md) {
  form {
    max-width: 90%;
  }
}
</style>
