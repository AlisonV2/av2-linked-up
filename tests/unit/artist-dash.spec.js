import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import ArtistDashboard from '@/components/artists/ArtistDashboard';
import AppTitle from '@/components/app/Title';
import AppButton from '@/components/app/Button';

const actions = {
  getArtistProfile: jest.fn(),
  setArtistThumbnail: jest.fn(),
  setArtistProfile: jest.fn(),
};

const state = {
  artistProfile: {
    artistProfile: {
      name: '',
      city: '',
      zip: '',
      shop: '',
      style: '',
      socialLink: '',
      insta: '',
      description: '',
    },
  },
};

const store = createStore({
  state,
  actions,
});

const $router = {
  push: jest.fn(),
};

const $route = {
  params: { id: '1' },
};

const config = {
  data() {
    return {
      previewImage: null,
      profile: {
        name: '',
        city: '',
        zip: '',
        shop: '',
        style: '',
        socialLink: '',
        insta: '',
        description: '',
      },
      showSuccessToast: false,
      showErrorToast: false,
      fileSelected: null,
      fileErr: null,
    };
  },
  methods: {
    setPreviewImage: jest.fn(),
  },
  global: {
    plugins: [store],
    components: {
      'app-title': AppTitle,
      'app-button': AppButton
    },
    mocks: {
      $router,
      $route,
    },
  },
};
/**
 * @module ArtistDashboardTest
 */
describe('ArtistDashboard.vue', () => {
  it('ArtistDashboard Snapshot', () => {
    const component = shallowMount(ArtistDashboard, config);
    expect(component.element).toMatchSnapshot();
  });

  it('Check if getArtistProfile action is called', async () => {
    const component = shallowMount(ArtistDashboard, config);
    ArtistDashboard.created.call(component.vm);
    await component.vm.$nextTick();
    expect(actions.getArtistProfile).toHaveBeenCalled;
  });

  it('Check if setArtistProfile action is called', async () => {
    const component = shallowMount(ArtistDashboard, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.setArtistProfile).toHaveBeenCalled;
  });

  it('Check if setArtistThumbnail action is called', async () => {
    const component = shallowMount(ArtistDashboard, config);
    component.find('.thumbnail-button').trigger('click');
    await component.vm.$nextTick();
    expect(actions.setArtistThumbnail).toHaveBeenCalled;
  });

  it('Check if toastError component is visible', async () => {
    const component = shallowMount(ArtistDashboard, {
      ...config,
      data() {
        return {
          showSuccessToast: false,
          showErrorToast: true,
        };
      },
    });
    const error = component.find('.toast-error');
    expect(error.exists()).toBe(true);
  });

  it('setPreviewImage should be called', () => {
    const component = shallowMount(ArtistDashboard, config);
    const file = component.find('input[type="file"]');
    file.trigger('change');
    expect(component.vm.setPreviewImage).toHaveBeenCalled;
  });
});
