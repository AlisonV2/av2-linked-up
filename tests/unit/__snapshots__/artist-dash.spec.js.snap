// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistDashboard.vue ArtistDashboard Snapshot 1`] = `
<div
  class="row"
>
  <app-title-stub />
  <!--v-if-->
  <!--v-if-->
  <div
    class="col-12 col-lg-4 mt-4"
  >
    <div
      class="mt-3"
    >
      <input
        data-test="file-upload"
        type="file"
      />
      <div
        class="error"
      />
    </div>
    <img
      class="img-fluid"
      height="250"
      src=""
      width="250"
    />
    <div
      class="btn-center mt-3 thumbnail-button"
    >
      <app-button
        mode="save-btn"
      >
        Save
      </app-button>
    </div>
    <!-- &lt;div class="error"&gt;{{ fileError }}&lt;/div&gt; -->
  </div>
  <form
    class="col-12 col-lg-8 mt-4"
  >
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Public Name"
        type="text"
      />
      <label>
        Public Name
      </label>
    </div>
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Description "
        type="textarea"
      />
      <label>
        Description
      </label>
    </div>
    <div
      class="input-group"
    >
      <div
        class="form-floating mb-3 col-lg-6"
      >
        <input
          class="form-control"
          placeholder="City"
          type="text"
        />
        <label>
          City
        </label>
      </div>
      <div
        class="form-floating mb-3 col-lg-6"
      >
        <input
          class="form-control"
          placeholder="Zip Code"
          type="text"
        />
        <label>
          Zip Code
        </label>
      </div>
    </div>
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Shop"
        type="text"
      />
      <label>
        Shop
      </label>
    </div>
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Social Media Link"
        type="text"
      />
      <label>
        Social Media Link
      </label>
    </div>
    <div
      class="form-floating mb-3"
    >
      <input
        class="form-control"
        placeholder="Instagram Name"
        type="text"
      />
      <label>
        Instagram Name
      </label>
    </div>
    <div>
      <select
        class="form-select"
      >
        <option
          disabled=""
          value=""
        >
          Choose your style
        </option>
        <option
          value="Black Work"
        >
          Black Work
        </option>
        <option
          value="New School"
        >
          New School
        </option>
        <option
          value="Old School"
        >
          Old School
        </option>
        <option
          value="Surrealism"
        >
          Surrealism
        </option>
      </select>
    </div>
    <div
      class="col-12 btn-right mt-3"
    >
      <div
        class="btn-group mb-3"
      >
        <app-button>
          Save
        </app-button>
      </div>
    </div>
  </form>
</div>
`;
